File: app/src/main/java/me/pwcong/jpstart/mvp/presenter/TranslateFragmentPresenterImpl.java
package me.pwcong.jpstart.mvp.presenter;

import me.pwcong.jpstart.App;
import me.pwcong.jpstart.R;
import me.pwcong.jpstart.manager.ClipboardManager;
import me.pwcong.jpstart.mvp.bean.BaiduTranslateBean;
import me.pwcong.jpstart.mvp.model.BaseModel;
import me.pwcong.jpstart.mvp.model.TranslateFragmentModelImpl;
import me.pwcong.jpstart.mvp.view.BaseView;
import me.pwcong.jpstart.network.baidu.BaiduTranslateApi;
import me.pwcong.jpstart.utils.StringUtils;
import rx.Subscriber;

/**
 * Created by Pwcong on 2016/10/2.
 */

public class TranslateFragmentPresenterImpl extends BasePresenter<BaseView.TranslateFragmentView> implements BasePresenter.TranslateFragmentPresenter {

    BaseModel.TranslateFragmentModel model;

    public TranslateFragmentPresenterImpl(BaseView.TranslateFragmentView view) {
        super(view);
        model=new TranslateFragmentModelImpl();
    }

    @Override
    public void initTranslateFragment() {

        view.setFromSpinner(model.getFromList());
        view.setToSpinner(model.getToList());

    }

    @Override
    public void checkFromLanguate(int from) {
        App.FROM_LAN=from;
    }

    @Override
    public void checkToLanguage(int to) {
        App.TO_LAN=to;
    }

    @Override
    public void checkImageViewClick(int id) {

        switch (id){

            case R.id.iv_src_copy:

                String srcText = view.getSrcText();
                if(!StringUtils.isNullOrEmpty(srcText)){
                    ClipboardManager.getInstance().setText("label",view.getSrcText());
                    view.showMsg(R.string.copy_successfully);
                }
                break;
            case R.id.iv_src_paste:
                view.setSrcEditText(ClipboardManager.getInstance().getText());
                break;
            case R.id.iv_src_clear:
                view.setSrcEditText("");
                break;
            case R.id.iv_dst_copy:

                String dstText = view.getDstText();
                if(!StringUtils.isNullOrEmpty(dstText)){
                    ClipboardManager.getInstance().setText("label",view.getDstText());
                    view.showMsg(R.string.copy_successfully);
                }
                break;
            case R.id.iv_dst_clear:
                view.setDstTextView("");
                break;
            default:break;
        }

    }

    @Override
    public void doTranslate() {

        String from;

        switch (App.FROM_LAN){

            case 0:from= BaiduTranslateApi.AUTO;break;
            case 1:from=BaiduTranslateApi.ZH;break;
            case 2:from=BaiduTranslateApi.EN;break;
            case 3:from=BaiduTranslateApi.JP;break;
            default:from=BaiduTranslateApi.AUTO;break;

        }

        String to;

        switch (App.TO_LAN){

            case 0:to=BaiduTranslateApi.ZH;break;
            case 1:to=BaiduTranslateApi.EN;break;
            case 2:to=BaiduTranslateApi.JP;break;
            default:to=BaiduTranslateApi.ZH;break;

        }

        model.translate(view.getSrcText(), from, to, new Subscriber<BaiduTranslateBean>() {
            @Override
            public void onCompleted() {
            }

            @Override
            public void onError(Throwable e) {
                view.showMsg(R.string.error_unknown);
                e.printStackTrace();
            }

            @Override
            public void onNext(BaiduTranslateBean baiduTranslateBean) {
                if(baiduTranslateBean.getError_code()==null){
                    view.setDstTextView(baiduTranslateBean.getTrans_result()[0].getDst());
                }else {
                    view.showMsg(baiduTranslateBean.getError_msg());
                }
            }
        });


    }
}

File: app/src/main/java/com/github/pwcong/jpstart/mvp/presenter/impl/TranslateFragmentPresenterImpl.kt
package com.github.pwcong.jpstart.mvp.presenter.impl

import com.github.pwcong.jpstart.App
import com.github.pwcong.jpstart.R
import com.github.pwcong.jpstart.manager.ClipboardManager
import com.github.pwcong.jpstart.mvp.bean.TranslateBean
import com.github.pwcong.jpstart.mvp.model.BaseModel
import com.github.pwcong.jpstart.mvp.model.impl.TranslateFragmentModelImpl
import com.github.pwcong.jpstart.mvp.presenter.BasePresenter
import com.github.pwcong.jpstart.mvp.presenter.BasePresenter.TranslateFragmentPresenter
import com.github.pwcong.jpstart.mvp.view.BaseView.TranslateFragmentView
import com.github.pwcong.jpstart.network.baidu.api.BaiduTranslateApi
import rx.Subscriber

class TranslateFragmentPresenterImpl(view: TranslateFragmentView) :
    BasePresenter<TranslateFragmentView>(view), TranslateFragmentPresenter {
    private val model: BaseModel.TranslateFragmentModel = TranslateFragmentModelImpl()

    override fun initTranslateFragment() {
        view.setFromSpinner(model.fromList)
        view.setToSpinner(model.toList)
    }

    override fun checkFromLanguage(from: Int) {
        App.FROM_LAN = from
    }

    override fun checkToLanguage(to: Int) {
        App.TO_LAN = to
    }

    override fun checkImageViewClick(id: Int) {
        when (id) {
            R.id.iv_src_copy -> {
                val srcText = view.srcText
                if (!srcText.isNullOrEmpty()) {
                    ClipboardManager.getInstance().setText("label", view.srcText)
                    view.showMsg(R.string.copy_successfully)
                }
            }

            R.id.iv_src_paste -> view.setSrcEditText(ClipboardManager.getInstance().text)
            R.id.iv_src_clear -> view.setSrcEditText("")
            R.id.iv_dst_copy -> {
                val dstText = view.dstText
                if (!dstText.isNullOrEmpty()) {
                    ClipboardManager.getInstance().setText("label", view.dstText)
                    view.showMsg(R.string.copy_successfully)
                }
            }

            R.id.iv_dst_clear -> view!!.setDstTextView("")
            else -> {}
        }
    }

    override fun doTranslate() {
        val srcText = view.srcText

        if (!srcText.isNullOrEmpty()) {
            val from: String = when (App.FROM_LAN) {
                0 -> BaiduTranslateApi.AUTO
                1 -> BaiduTranslateApi.ZH
                2 -> BaiduTranslateApi.EN
                3 -> BaiduTranslateApi.JP
                else -> BaiduTranslateApi.AUTO
            }

            val to: String = when (App.TO_LAN) {
                0 -> BaiduTranslateApi.ZH
                1 -> BaiduTranslateApi.EN
                2 -> BaiduTranslateApi.JP
                else -> BaiduTranslateApi.ZH
            }
            model.translate(view.srcText, from, to, object : Subscriber<TranslateBean>() {
                override fun onCompleted() {
                }

                override fun onError(e: Throwable) {
                    view.showMsg(R.string.translate_error)
                    e.printStackTrace()
                }

                override fun onNext(translateBean: TranslateBean) {
                    if (translateBean.code == 200) {
                        view.setDstTextView(translateBean.result)
                    } else {
                        view.showMsg(R.string.translate_error)
                    }
                }
            })
        }
    }
}
